### PR TITLE
fix(Transfer): Dropdown is hidden & `showSizeChanger` not work

### LINKED ISSUE #41790: Transfer 穿梭框分页 SizeChanger 无效且缺少左右 padding 样式
### Reproduction link

[Edit on CodeSandbox](https://codesandbox.io/s/fen-ye-pagesizeqie-huan-wu-xiao-antd-5-4-2-forked-e2vksi)

### Steps to reproduce

1. 为 Transfer 组件设置 `pagination={{ showSizeChanger: true, simple: false }}`

### What is expected?

1. `pagination` 右侧距离 `Transfer` 右侧边框有 `padding`
2. `SizeChanger` 可以正确切换每页数量

### What is actually happening?

1. 发现 `pagination` 右侧紧贴 `Transfer` 右侧边框
2. 发现 `SizeChanger` 打不开或打开后无法选择任何选项

| Environment | Info |
| --- | --- |
| antd | 5.4.2 |
| React | 18.2.0 |
| System | Windows 11 |
| Browser | Chrome 112 |

---
1. 在某些情况下 `SizeChanger` 可以打开但选择具体数量后并未改变
2. `PaginationType` 中没有类似 `onPageSizeChange` 的函数也就无法控制 `pageSize` 的值
3. `SizeChanger` 打不开在 `react-grid-layout` 中单独使用 `Pagination` 时也会出现，暂时无法复现

### PR DESCRIPTION
**Pull request by Yuiai01**
[[中文版模板 / Chinese template](https://github.com/ant-design/ant-design/blob/master/.github/PULL_REQUEST_TEMPLATE/pr_cn.md?plain=1)]

### 🤔 This is a ...

- [ ] New feature
- [x] Bug fix
- [ ] Site / documentation update
- [ ] Demo update
- [ ] Component style update
- [ ] TypeScript definition update
- [ ] Bundle size optimization
- [ ] Performance optimization
- [ ] Enhancement feature
- [ ] Internationalization
- [ ] Refactoring
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Workflow
- [ ] Other (about what?)

### 🔗 Related issue link

close https://github.com/ant-design/ant-design/issues/41790
### 💡 Background and solution

在当前的 `Transfer` 组件中，如果设置了 `pagination={{ showSizeChanger: true, simple: false }}` ，那么会出现切换分页数量的下拉按钮被隐藏的情况：
image
解决被隐藏问题并添加右边距（之前贴边）后，如下图：
image
根据 https://github.com/ant-design/ant-design/issues/41790#issuecomment-1507831602，发现 `Transfer` 组件内部的 `Pagination` 没有绑定 `showSizeChanger` 方法，因此添加上对应方法。

### 📝 Changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English | fix Pagination Dropdown is hidden & `showSizeChanger` not work           |
| 🇨🇳 Chinese | 修复分页下拉按钮被隐藏以及 `showSizeChanger` 方法无效          |

### ☑️ Self-Check before Merge

⚠️ Please check all items below before requesting a reviewing. ⚠️

- [x] Doc is updated/provided or not needed
- [x] Demo is updated/provided or not needed
- [x] TypeScript definition is updated/provided or not needed
- [x] Changelog is provided or not needed

---

### <samp>🤖 Generated by Copilot at ff72fad</samp>

Added pagination support for the `Transfer` component, allowing users to select and change the page size of the transfer list. Updated the test cases, the `TransferList` and `ListBody` components, and the style file to handle the pagination functionality.

### <samp>🤖 Generated by Copilot at ff72fad</samp>

*  Add `pageSize` and `onPageSizeChange` props to `TransferList` and `ListBody` components to enable changing the number of items per page in the transfer component ([link](https://github.com/ant-design/ant-design/pull/41906/files?diff=unified&w=0#diff-c6feaf8f7f09d7e3901905a89bc11ecd461a7340d5a8e8ea4dbc994a77bdc415L113-R113), [link](https://github.com/ant-design/ant-design/pull/41906/files?diff=unified&w=0#diff-c6feaf8f7f09d7e3901905a89bc11ecd461a7340d5a8e8ea4dbc994a77bdc415R204-R207), [link](https://github.com/ant-design/ant-design/pull/41906/files?diff=unified&w=0#diff-c6feaf8f7f09d7e3901905a89bc11ecd461a7340d5a8e8ea4dbc994a77bdc415R232), [link](https://github.com/ant-design/ant-design/pull/41906/files?diff=unified&w=0#diff-2234478f9c4d5f6c74fa4089e1f1ac4f7813fc5d971297a37959055c8eb85d6cL17-R21), [link](https://github.com/ant-design/ant-design/pull/41906/files?diff=unified&w=0#diff-2234478f9c4d5f6c74fa4089e1f1ac4f7813fc5d971297a37959055c8eb85d6cL56-R72), [link](https://github.com/ant-design/ant-design/pull/41906/files?diff=unified&w=0#diff-2234478f9c4d5f6c74fa4089e1f1ac4f7813fc5d971297a37959055c8eb85d6cR113-R116))
*  Modify `parsePagination` function to accept an optional `pageSize` parameter and use it to slice the render items according to the current page number and page size ([link](https://github.com/ant-design/ant-design/pull/41906/files?diff=unified&w=0#diff-2234478f9c4d5f6c74fa4089e1f1ac4f7813fc5d971297a37959055c8eb85d6cL25-R27), [link](https://github.com/ant-design/ant-design/pull/41906/files?diff=unified&w=0#diff-2234478f9c4d5f6c74fa4089e1f1ac4f7813fc5d971297a37959055c8eb85d6cL84-R87), [link](https://github.com/ant-design/ant-design/pull/41906/files?diff=unified&w=0#diff-2234478f9c4d5f6c74fa4089e1f1ac4f7813fc5d971297a37959055c8eb85d6cL92-R99))
*  Add `onShowSizeChange` prop to `Pagination` component to trigger the `onPageSizeChange` callback and reset the current page number when the user selects a different page size ([link](https://github.com/ant-design/ant-design/pull/41906/files?diff=unified&w=0#diff-2234478f9c4d5f6c74fa4089e1f1ac4f7813fc5d971297a37959055c8eb85d6cR113-R116))
*  Add `generateData` function to `index.test.tsx` to create mock data items for testing the transfer component with pagination ([link](https://github.com/ant-design/ant-design/pull/41906/files?diff=unified&w=0#diff-4c216b8afeb5629f3befe0a7945c5f6f492158195855764e91524049a0462296R76-R88))
*  Add test case for changing the page size in the transfer component and verify the correct number of items per page ([link](https://github.com/ant-design/ant-design/pull/41906/files?diff=unified&w=0#diff-4c216b8afeb5629f3befe0a7945c5f6f492158195855764e91524049a0462296R587-R597))
*  Add `antCls` parameter to `genTransferListStyle` function in `style/index.ts` to access the ant design class name for the transfer list component ([link](https://github.com/ant-design/ant-design/pull/41906/files?diff=unified&w=0#diff-d9d90f3711c4b10ecac467dc5d915fd5e15d8398916ed340a5ca70968394f2cbR106))
*  Add `minHeight` property to the transfer list style to fix a flexbox bug that causes the list body to overflow the list container when the page size is larger than the number of items ([link](https://github.com/ant-design/ant-design/pull/41906/files?diff=unified&w=0#diff-d9d90f3711c4b10ecac467dc5d915fd5e15d8398916ed340a5ca70968394f2cbL174-R177))
*  Add `paddingInlineEnd` property to the pagination options style to add some spacing between the page size changer and the pagination buttons ([link](https://github.com/ant-design/ant-design/pull/41906/files?diff=unified&w=0#diff-d9d90f3711c4b10ecac467dc5d915fd5e15d8398916ed340a5ca70968394f2cbR264-R267))

### CODE REVIEW

File: components/transfer/ListBody.tsx [hunk 1]
```diff
@@ -14,15 +14,17 @@ export interface TransferListBodyProps<RecordType> extends PartialTransferListPr
   filteredItems: RecordType[];
   filteredRenderItems: RenderedItem<RecordType>[];
   selectedKeys: string[];
+  pageSize: number;
+  onPageSizeChange: (size: number) => void;
 }
 
-const parsePagination = (pagination?: PaginationType) => {
+const parsePagination = (pagination?: PaginationType, pageSize?: number) => {
```
Comment (MadCcc) [inline components/transfer/ListBody.tsx:21]: pageSize 应该包含在 pagination 里？

Comment (Yuiai01) [reply components/transfer/ListBody.tsx:21]: 放在里面的话会出现我切换 pageSize 后页面数据不会刷新

Comment (MadCcc) [reply components/transfer/ListBody.tsx:21]: 这很奇怪，不应该单独维护一个 pageSize，可以看看什么原因导致

Comment (Yuiai01) [reply components/transfer/ListBody.tsx:21]: > 这很奇怪，不应该单独维护一个 pageSize，可以看看什么原因导致

ListBody 内部通过 ref 将过滤好的数据传递给外层，如果设置 `showSizeChanger` 后在 ListBody 内部维护 pageSize 那么就会出现我切换完 pageSize 后页面数据不会变化，因为没有触发外层组件的刷新。因此我将 `pageSize` 提取出来单独维护了

**Comment by afc163** [comment]
any progress?

**Comment by Yuiai01** [comment]
> any progress?

好了~

File: components/transfer/ListBody.tsx [hunk 2]
```diff
@@ -52,21 +44,23 @@ const TransferListBody: React.ForwardRefRenderFunction<
...
+  const convertPagination = typeof pagination === 'object' ? pagination : {};
 
   const [current, setCurrent] = React.useState<number>(1);
+  const [pageSize, setPageSize] = React.useState<number>(convertPagination?.pageSize ?? 10);
 
   React.useEffect(() => {
-    const mergedPagination = parsePagination(pagination);
-    if (mergedPagination) {
+    const mergedPagination = parsePagination({ ...convertPagination, pageSize });
+    if (pagination) {
```
Comment (MadCcc) [inline components/transfer/ListBody.tsx:59]:
```suggestion
    if (mergedPagination) {
```
这里逻辑应该不变？

File: components/transfer/ListBody.tsx [hunk 3]
```diff
@@ -52,21 +44,23 @@ const TransferListBody: React.ForwardRefRenderFunction<
...
       const maxPageCount = Math.ceil(filteredRenderItems.length / mergedPagination.pageSize!);
       setCurrent(Math.min(current, maxPageCount));
     }
-  }, [filteredRenderItems, pagination]);
+  }, [filteredRenderItems, pagination, pageSize]);
```
Comment (MadCcc) [inline components/transfer/ListBody.tsx:63]:
```suggestion
  }, [filteredRenderItems, convertPagination, pageSize]);
```

File: components/transfer/ListBody.tsx [hunk 4]
```diff
@@ -52,21 +44,23 @@ const TransferListBody: React.ForwardRefRenderFunction<
...
-    pagination,
+    pagination = null,
     onScroll,
     onItemSelect,
     onItemRemove,
   } = props;
+  const convertPagination = typeof pagination === 'object' ? pagination : {};
 
   const [current, setCurrent] = React.useState<number>(1);
+  const [pageSize, setPageSize] = React.useState<number>(convertPagination?.pageSize ?? 10);
```
Comment (MadCcc) [inline components/transfer/ListBody.tsx:55]: 如果 pagination 里传进来的 pageSize 改变了是否应该响应？

File: components/transfer/ListBody.tsx [hunk 5]
```diff
@@ -80,20 +74,25 @@ const TransferListBody: React.ForwardRefRenderFunction<
...
 
+  const onSizeChange = (cur: number, size: number) => {
+    setCurrent(cur);
+    setPageSize(size);
+  };
+
   const memoizedItems = React.useMemo<RenderedItem<RecordType>[]>(() => {
-    const mergedPagination = parsePagination(pagination);
-    const displayItems = mergedPagination
+    const mergedPagination = parsePagination({ ...convertPagination, pageSize });
```
Comment (MadCcc) [inline components/transfer/ListBody.tsx:83]: 感觉这一行写了很多遍，可以单独写一个 useMemo 吧

File: components/transfer/ListBody.tsx (hunk 3 above)
Comment (Yuiai01) [reply components/transfer/ListBody.tsx]: 修改掉了

File: components/transfer/ListBody.tsx [hunk 6]
```diff
@@ -1,40 +1,33 @@
...
 export interface TransferListBodyProps<RecordType> extends PartialTransferListProps<RecordType> {
   filteredItems: RecordType[];
   filteredRenderItems: RenderedItem<RecordType>[];
   selectedKeys: string[];
 }
 
-const parsePagination = (pagination?: PaginationType) => {
-  if (!pagination) {
-    return null;
-  }
```
Comment (MadCcc) [inline components/transfer/ListBody.tsx:22]: 这里判空还是得有？

File: components/transfer/ListBody.tsx [hunk 7]
```diff
@@ -52,21 +45,29 @@ const TransferListBody: React.ForwardRefRenderFunction<
     selectedKeys,
     disabled: globalDisabled,
     showRemove,
-    pagination,
+    pagination = null,
     onScroll,
     onItemSelect,
     onItemRemove,
   } = props;
+  const convertPagination = typeof pagination === 'object' ? pagination : {};
```
Comment (MadCcc) [inline components/transfer/ListBody.tsx:53]: 感觉这里不等价，pagination 可以是 `false` 的话就和 `true` 一样处理了。
并且 null 也是 `object`

Comment (MadCcc) [reply components/transfer/ListBody.tsx:53]: 我是觉得应该一并收到 mergedPagination 这个 memo 里处理好点？

**Comment by MadCcc** [comment]
受控还是要用 `useMergedState`，props 里的作为 value，最终在下面使用的都是 pageSize 这个 state 而不是 mergedPagination 里的 pageSize。

**Comment by YMNNs** [comment]
@MadCcc 可以考虑合并一下这个PR吗？

**Comment by afc163** [comment]
@YMNNs ci 挂了

**Comment by Yuiai01** [comment]
> @MadCcc 可以考虑合并一下这个PR吗？

我这边跟进一下～

**Review by MadCcc — APPROVED** [review]

**Comment by MadCcc** [comment]
argos 截图好像变了
https://app.argos-ci.com/ant-design/ant-design/builds/9547/50739793

**Comment by Yuiai01** [comment]
> argos 截图好像变了 https://app.argos-ci.com/ant-design/ant-design/builds/9547/50739793

好，我康康

**Comment by afc163** [comment]
图片

这个样式这么乱，感觉不太对。

**Comment by Yuiai01** [comment]
> 图片
> 这个样式这么乱，感觉不太对。

这个样式是自定义出来的，可能自定义的参数与普通的组件样式差异过大所以看起来比较乱。
image

**Pull request merged**